src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlStatistics.cs

```diff
@@ -335,4 +337,16 @@ void ICollection.CopyTo(Array array, int arrayIndex)
             }
         }
     }
+
+    // This is a ref struct to prevent it being included in async closures accidentally. 
+    // Async functions should manage the timer directly using the Start and Stop method
+    // in their invoke and completion functions
+    internal readonly ref struct ValueSqlStatisticsScope // : IDisposable // ref structs cannot implement interfaces but the compiler will use pattern matching to allow use of using on them
+    {
+        private readonly SqlStatistics _statistics;
+
+        public ValueSqlStatisticsScope(SqlStatistics statistics) => _statistics = SqlStatistics.StartTimer(statistics);
+
+        public void Dispose() => SqlStatistics.StopTimer(_statistics);
```

Comment:
Could you implement the dispose pattern here, i.e. `public void Dispose() => Dispose(true)`?

Reply:
Yes but it would have no value. A ref struct cannot ever have a finalizer or be present on the finalization heap so there is no need to distinguish between disposal and finalization. It also holds no native resources that would require free-ing if dropped to GC so it's no a useful pattern in general.

Reply:
https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=net-5.0#implementing-idisposable